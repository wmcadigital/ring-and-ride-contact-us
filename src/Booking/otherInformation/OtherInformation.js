import { useFormState } from "react-final-form";

import FormSection from "../../common/FormSection";
import RadioGroup from "../../common/RadioGroup";
import Question from "../../common/Question";
import RadioButton from "../../common/RadioButton";
import TextArea from "../../common/TextArea";
import FieldError from "../../common/FieldError";
import ProgressIndicator from "../../common/ProgressIndicator";
import getSectionPositionInfo from "../getSectionPosition";
import { required } from "../../common/validation";

const OtherInformation = () => {
  const stateApi = useFormState();
  const formValues = stateApi.values;

  const error = stateApi.submitFailed
    ? stateApi.errors?.otherInformation
    : null;

  const otherInformationDetailsError = stateApi.submitFailed
    ? stateApi.errors?.otherInformationDetails
    : null;

  return (
    <FormSection>
      <ProgressIndicator
        sectionPosition={getSectionPositionInfo("4", stateApi)}
        sectionName="Any other information"
      />
      <Question text="Is there any other information you think we should know about this booking?" />
      <RadioGroup error={error}>
        <FieldError text={error} />
        <RadioButton
          key={1}
          label="Yes"
          validation={required}
          value="yes"
          fieldName="otherInformation"
        />
        {formValues.otherInformation === "yes" ? (
          <TextArea
            label="Additional requirements"
            fieldName="otherInformationDetails"
            validation={required}
            error={otherInformationDetailsError}
            rows={8}
          />
        ) : null}
        <RadioButton
          key={2}
          label="No"
          validation={required}
          value="no"
          fieldName="otherInformation"
        />
      </RadioGroup>
    </FormSection>
  );
};

export default OtherInformation;
