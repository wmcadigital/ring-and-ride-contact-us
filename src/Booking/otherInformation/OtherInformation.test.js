import { useFormState } from "react-final-form";
import { createRenderer } from "react-test-renderer/shallow";

import OtherInformation from "./OtherInformation";

jest.mock("react-final-form");

describe("OtherInformation - snapshots", () => {
  it("individual is filling out form for themselves", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "mySelf",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<OtherInformation />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("individual is filling out form for someone else", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "behalfSomeone",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<OtherInformation />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("individual is filling out form for a group", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "behalfGroup",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<OtherInformation />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an error", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "mySelf",
      },
      submitFailed: true,
      errors: { otherInformation: "Required" },
    }));
    const renderer = createRenderer();

    renderer.render(<OtherInformation />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("shows text area for other information if yes is selected", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "mySelf",
        otherInformation: "yes",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<OtherInformation />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("shows error if text area not filled in on submission", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "mySelf",
        otherInformation: "yes",
      },
      submitFailed: true,
      errors: { otherInformationDetails: "Required" },
    }));
    const renderer = createRenderer();

    renderer.render(<OtherInformation />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
