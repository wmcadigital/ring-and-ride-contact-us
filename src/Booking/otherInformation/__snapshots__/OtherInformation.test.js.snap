// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OtherInformation - snapshots individual is filling out form for a group 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Any other information"
    sectionPosition="Section 5 of 5"
  />
  <Question
    text="Is there any other information you think we should know about this booking?"
  />
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="otherInformation"
      label="Yes"
      onClickCallback={[Function]}
      validation={[Function]}
      value="yes"
    />
    <RadioButton
      fieldName="otherInformation"
      label="No"
      onClickCallback={[Function]}
      validation={[Function]}
      value="no"
    />
  </RadioGroup>
</FormSection>
`;

exports[`OtherInformation - snapshots individual is filling out form for someone else 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Any other information"
    sectionPosition="Section 4 of 4"
  />
  <Question
    text="Is there any other information you think we should know about this booking?"
  />
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="otherInformation"
      label="Yes"
      onClickCallback={[Function]}
      validation={[Function]}
      value="yes"
    />
    <RadioButton
      fieldName="otherInformation"
      label="No"
      onClickCallback={[Function]}
      validation={[Function]}
      value="no"
    />
  </RadioGroup>
</FormSection>
`;

exports[`OtherInformation - snapshots individual is filling out form for themselves 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Any other information"
    sectionPosition="Section 4 of 4"
  />
  <Question
    text="Is there any other information you think we should know about this booking?"
  />
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="otherInformation"
      label="Yes"
      onClickCallback={[Function]}
      validation={[Function]}
      value="yes"
    />
    <RadioButton
      fieldName="otherInformation"
      label="No"
      onClickCallback={[Function]}
      validation={[Function]}
      value="no"
    />
  </RadioGroup>
</FormSection>
`;

exports[`OtherInformation - snapshots renders with an error 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Any other information"
    sectionPosition="Section 4 of 4"
  />
  <Question
    text="Is there any other information you think we should know about this booking?"
  />
  <RadioGroup
    error="Required"
  >
    <FieldError
      text="Required"
    />
    <RadioButton
      fieldName="otherInformation"
      label="Yes"
      onClickCallback={[Function]}
      validation={[Function]}
      value="yes"
    />
    <RadioButton
      fieldName="otherInformation"
      label="No"
      onClickCallback={[Function]}
      validation={[Function]}
      value="no"
    />
  </RadioGroup>
</FormSection>
`;

exports[`OtherInformation - snapshots shows error if text area not filled in on submission 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Any other information"
    sectionPosition="Section 4 of 4"
  />
  <Question
    text="Is there any other information you think we should know about this booking?"
  />
  <RadioGroup>
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="otherInformation"
      label="Yes"
      onClickCallback={[Function]}
      validation={[Function]}
      value="yes"
    />
    <TextArea
      error="Required"
      fieldName="otherInformationDetails"
      label="Additional requirements"
      rows={8}
      validation={[Function]}
    />
    <RadioButton
      fieldName="otherInformation"
      label="No"
      onClickCallback={[Function]}
      validation={[Function]}
      value="no"
    />
  </RadioGroup>
</FormSection>
`;

exports[`OtherInformation - snapshots shows text area for other information if yes is selected 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Any other information"
    sectionPosition="Section 4 of 4"
  />
  <Question
    text="Is there any other information you think we should know about this booking?"
  />
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="otherInformation"
      label="Yes"
      onClickCallback={[Function]}
      validation={[Function]}
      value="yes"
    />
    <TextArea
      error={null}
      fieldName="otherInformationDetails"
      label="Additional requirements"
      rows={8}
      validation={[Function]}
    />
    <RadioButton
      fieldName="otherInformation"
      label="No"
      onClickCallback={[Function]}
      validation={[Function]}
      value="no"
    />
  </RadioGroup>
</FormSection>
`;
